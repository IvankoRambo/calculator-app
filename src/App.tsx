import { useState, MouseEvent } from 'react';
import Display from './components/Display';
import Panel from './components/Panel';
import { updateCalculation } from './utils/callApi';
import isFractionAllowed from './utils/isFractionAllowed';

function App() {
  const [equation, setEquation] = useState<string>('');
  const [result, setResult] = useState<string>('');
  const [disabled, setDisabled] = useState<boolean>(false);

  const operators = ['/', '*', '-', '+', '.'];

  const handleequation = async (event: MouseEvent) => {
    const value = (event.target as HTMLButtonElement).textContent || '';

    if (
      value &&
      (
        value !== '0' ||
        equation
      ) &&
      (
        !operators.includes(value) ||
        (
          equation &&
          !operators.includes(equation.slice(-1)) &&
          !(value === '.' && !isFractionAllowed(equation))
        )
      )
    ) {
      setEquation(equation + value);

      if (!operators.includes(value)) {
        const { success, result: updatedResult } = await updateCalculation(equation + value);

        if (success) {
          setResult(updatedResult);
        }
      }
    }
  };

  const handleSummarisation = async () => {
    setDisabled(true);

    const { success, result: updatedResult } = await updateCalculation(equation);

    setDisabled(false);

    if (success && result != null) {
      setEquation(updatedResult.toString());
    }
  };

  const handleDeletion = async () => {
    if (equation) {
      const updatedEquation = equation.slice(0, -1);

      setEquation(updatedEquation);

      if (!operators.includes(updatedEquation.slice(-1))) {
        const { success, result: updatedResult } = await updateCalculation(updatedEquation);

        if (success) {
          setResult(updatedResult);
        }
      }
    }
  };

  const handleClearing = async () => {
    if (equation) {
      const updatedEquation = '';

      setEquation(updatedEquation);

      const { success, result: updatedResult } = await updateCalculation(updatedEquation);

      if (success) {
        setResult(updatedResult);
      }
    }
  };

  return (
    <div className="App">
      <div className="calculator">
        <Display equation={equation} result={result} />
        <Panel
          equationHandler={handleequation}
          resultHandler={handleSummarisation}
          deletionHandler={handleDeletion}
          clearingHandler={handleClearing}
          disabled={disabled}
        />
      </div>
    </div>
  );
}

export default App;
