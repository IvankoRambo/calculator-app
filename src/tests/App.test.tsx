import { render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import App from '../App';

const getApiMock = (success: boolean, result?: string) => {
  return rest.put('/api/calculate', (req, res, ctx) => {
    return res(ctx.json({ success, result }));
  });
};

const server = setupServer(getApiMock(true));

describe('App Component', () => {
  beforeAll(() => server.listen());

  afterEach(() => server.resetHandlers());

  afterAll(() => server.close());

  it('should print number on digit click both for equation and preview part', async () => {
    server.use(getApiMock(true, '7'));

    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '7' }));

    expect(await screen.findByText(/7 =/)).toBeInTheDocument();
    expect(screen.getByTestId('equation')).toHaveTextContent('7');
  });

  it('should calculate equations correctly', async () => {
    const equation = '2*2/4+6-1';
    const expectedResult = '6';

    server.use(getApiMock(true, '6'));

    const { asFragment } = render(<App />);

    userEvent.click(screen.getByRole('button', { name: '2' }));
    userEvent.click(screen.getByRole('button', { name: '*' }));
    userEvent.click(screen.getByRole('button', { name: '2' }));
    userEvent.click(screen.getByRole('button', { name: '/' }));
    userEvent.click(screen.getByRole('button', { name: '4' }));
    userEvent.click(screen.getByRole('button', { name: '+' }));
    userEvent.click(screen.getByRole('button', { name: '6' }));
    userEvent.click(screen.getByRole('button', { name: '-' }));
    userEvent.click(screen.getByRole('button', { name: '1' }));

    expect(await screen.findByText(new RegExp(`${expectedResult} =`))).toBeInTheDocument();
    expect(screen.getByTestId('equation')).toHaveTextContent(equation);

    userEvent.click(screen.getByRole('button', { name: '=' }));

    expect(await screen.findByTestId('equation')).toHaveTextContent(expectedResult);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not allow adding additional zero to original equation', () => {
    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '0' }));
    userEvent.click(screen.getByRole('button', { name: '0' }));

    expect(screen.getByTestId('equation')).toHaveTextContent('0');
  });

  it('should not allow put operator as a first symbol of equation', () => {
    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '+' }));

    expect(screen.getByTestId('equation')).toHaveTextContent('0');
  });

  it('should not allow put operator after another operator', () => {
    const equation = '3-1+';

    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '3' }));
    userEvent.click(screen.getByRole('button', { name: '-' }));
    userEvent.click(screen.getByRole('button', { name: '1' }));
    userEvent.click(screen.getByRole('button', { name: '+' }));
    userEvent.click(screen.getByRole('button', { name: '/' }));

    expect(screen.getByTestId('equation')).toHaveTextContent(equation);
  });

  it('should not allow put double fraction operator for one number', () => {
    const equation = '34.35';

    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '3' }));
    userEvent.click(screen.getByRole('button', { name: '4' }));
    userEvent.click(screen.getByRole('button', { name: '.' }));
    userEvent.click(screen.getByRole('button', { name: '3' }));
    userEvent.click(screen.getByRole('button', { name: '5' }));
    userEvent.click(screen.getByRole('button', { name: '.' }));

    expect(screen.getByTestId('equation')).toHaveTextContent(equation);
  });

  it('should not allow put double fraction operator for number in long equations', () => {
    const equation = '25/5+1.1+2+11.2';

    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '2' }));
    userEvent.click(screen.getByRole('button', { name: '5' }));
    userEvent.click(screen.getByRole('button', { name: '/' }));
    userEvent.click(screen.getByRole('button', { name: '5' }));
    userEvent.click(screen.getByRole('button', { name: '+' }));
    userEvent.click(screen.getByRole('button', { name: '1' }));
    userEvent.click(screen.getByRole('button', { name: '.' }));
    userEvent.click(screen.getByRole('button', { name: '1' }));
    userEvent.click(screen.getByRole('button', { name: '+' }));
    userEvent.click(screen.getByRole('button', { name: '2' }));
    userEvent.click(screen.getByRole('button', { name: '+' }));
    userEvent.click(screen.getByRole('button', { name: '1' }));
    userEvent.click(screen.getByRole('button', { name: '1' }));
    userEvent.click(screen.getByRole('button', { name: '.' }));
    userEvent.click(screen.getByRole('button', { name: '2' }));
    userEvent.click(screen.getByRole('button', { name: '.' }));

    expect(screen.getByTestId('equation')).toHaveTextContent(equation);
  });

  it('should erase one symbol per one click on DEL button', async () => {
    const expectedEquation = '1+1';
    const expectedResult = '2';

    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '1' }));

    server.use(getApiMock(true, '1'));

    userEvent.click(screen.getByRole('button', { name: '+' }));
    userEvent.click(screen.getByRole('button', { name: '1' }));

    server.use(getApiMock(true, '2'));

    userEvent.click(screen.getByRole('button', { name: '-' }));
    userEvent.click(screen.getByRole('button', { name: '2' }));

    server.use(getApiMock(true, '0'));

    userEvent.click(screen.getByRole('button', { name: 'DEL' }));
    userEvent.click(screen.getByRole('button', { name: 'DEL' }));

    server.use(getApiMock(true, expectedResult));

    expect(await screen.findByText(new RegExp(`${expectedResult} =`))).toBeInTheDocument();
    expect(screen.getByTestId('equation')).toHaveTextContent(expectedEquation);
  });

  it('should erase all equation and preview result, when C button is clicked', async () => {
    render(<App />);

    userEvent.click(screen.getByRole('button', { name: '2' }));
    userEvent.click(screen.getByRole('button', { name: '+' }));
    userEvent.click(screen.getByRole('button', { name: '2' }));

    server.use(getApiMock(true, '4'));

    expect(await screen.findByText(/4 =/)).toBeInTheDocument();
    expect(screen.getByTestId('equation')).toHaveTextContent('2+2');

    userEvent.click(screen.getByRole('button', { name: 'C' }));

    server.use(getApiMock(true, ''));

    await waitFor(() => {
      expect(screen.queryByText(/4 =/)).not.toBeInTheDocument();
    });
    expect(screen.getByTestId('equation')).toHaveTextContent('0');
  });
});
