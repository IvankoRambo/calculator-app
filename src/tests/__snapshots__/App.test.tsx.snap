// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should calculate equations correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="calculator"
    >
      <div
        class="input"
      >
        <span
          class="result"
        >
          6 =
        </span>
        Â 
        <span
          class="equation"
          data-testid="equation"
        >
          2*2/4+6-1
        </span>
      </div>
      <div
        class="operators"
      >
        <button
          class=""
          disabled=""
        >
          /
        </button>
        <button
          class=""
          disabled=""
        >
          *
        </button>
        <button
          class=""
          disabled=""
        >
          -
        </button>
        <button
          class=""
          disabled=""
        >
          +
        </button>
        <button
          class=""
          disabled=""
        >
          DEL
        </button>
        <button
          class=""
          disabled=""
        >
          C
        </button>
      </div>
      <div
        class="digits"
      >
        <button
          class=""
          disabled=""
        >
          1
        </button>
        <button
          class=""
          disabled=""
        >
          2
        </button>
        <button
          class=""
          disabled=""
        >
          3
        </button>
        <button
          class=""
          disabled=""
        >
          4
        </button>
        <button
          class=""
          disabled=""
        >
          5
        </button>
        <button
          class=""
          disabled=""
        >
          6
        </button>
        <button
          class=""
          disabled=""
        >
          7
        </button>
        <button
          class=""
          disabled=""
        >
          8
        </button>
        <button
          class=""
          disabled=""
        >
          9
        </button>
        <button
          class=""
          disabled=""
        >
          0
        </button>
        <button
          class=""
          disabled=""
        >
          .
        </button>
        <button
          class="equal"
          disabled=""
        >
          =
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
